url.bs

```diff
@@ -2476,39 +2476,81 @@ background information. [[!HTML]]
- <a>percent-encoded bytes</a> replaced with code points resulting from
- <a>percent decoding</a> those sequences converted to bytes, unless that renders those
- sequences invisible.
+<p>In a space-constrained display, URLs should be elided carefully to avoid misleading the user when
+making a security decision:
+
+<ul>
+ <li><p>Ensure that at least the <a for=host>registrable domain</a> can be shown when the URL is
+ rendered (to avoid showing, e.g., <code>...examplecorp.com</code> when loading
+ <code>https://not-really-examplecorp.com/</code>).
+
+ <li><p>When the full <a for=url>host</a> cannot be rendered, elide domain labels starting from the
+ front. For example, <code>examplecorp.com.evil.com</code> should be elided as
+ <code>...com.evil.com</code>, not <code>examplecorp.com...</code>. (Note that bidirectional text
+ means that the front of the string may not appear at the left.)
 </ul>
 
-<p>For the purposes of bidirectional text it should be rendered as if it were in a
-left-to-right embedding. [[!BIDI]]
+<h4 id=url-rendering-i18n>Internationalization and special characters</h4>
+
+<p>International domain names (IDNs), special characters, and bidirectional text should be handled
+with care to prevent spoofing:
+
+<ul>
+ <li><p>A <a for=/>URL</a>'s <a for=url>host</a> should be rendered using <a>domain to
+ Unicode</a>.
 
-<p class="note no-backref">Unfortunately, as rendered <a for=/>URLs</a> are strings and can appear
-anywhere, a specific bidirectional algorithm for rendered <a for=/>URLs</a> would not see wide
-adoption. Bidirectional text interacts with the parts of a <a for=/>URL</a> in ways that can cause
-the rendering to be different from the model. Users of bidirectional languages are thus cautioned
-that this is to be expected, particularly in plain text environments.
+ <p class="note no-backref">Note that non-ASCII characters can be used in <a
+ href="http://unicode.org/faq/idn.html#26">homograph</a> spoofing attacks. Consider detecting <a
+ href="http://www.unicode.org/reports/tr39/#Confusable_Detection">confusable characters</a> or
+ warning when they are in use.
 
-<p>Due to the confusion that can arise between a <a for=/>URL</a>'s <a for=url>host</a>
-and <a for=url>path</a> with bidirectional text, browsers are encouraged to only render a
-<a for=/>URL</a>'s <a for=url>host</a> in places where it is important for users to
-distinguish between the two. E.g., users are expected to make trust decisions based on a
-<a for=/>URL</a>'s <a for=url>host</a> rendered in the address bar.
+ <li><p>URLs are particularly prone to confusion between host and path when they contain
+ bidirectional text, so in this case it is particularly advisable to only render a URL’s <a
+ for=url>host</a>. Other parts of the <a for=/>URL</a>, if rendered, should have their sequences of
+ <a>percent-encoded bytes</a> replaced with code points resulting from <a>percent decoding</a> those
```

Comment:
I probably haven't thought about this enough, but I've lost track of how replacing %-encoded bytes with their decodings help disambiguate between the host and path. An example might help?

Reply:
This bit was in the original section, so I'm not sure what the original motivation was. My motivation for keeping it is that we should avoid having URLs that are so syntax-y that users don't even bother trying to understand them. I prepended "For readability," to this sentence to try to make that clearer. Do you think that helps? (I also added a note that some characters may not be decoded if they present spoofing risks, e.g. Unicode lock.)

Reply:
I guess it depends on how much we expect the path/query/fragment to show up going forward. It's definitely a little fragile to decode as the browser will later have to encode it again when going to the network. Differences there could result in issues (and have).

Reply:
FWIW I don't expect path/query/fragment from all URL display surfaces. They'll still be displayed when editing URLs in the address bar, and in secondary UI like bookmarks and history, etc. It probably makes sense to continue advising to decode in contexts like those.

Reply:
It might help to put "Other parts of the URL …" in a separate `<li>` since it's a good idea but doesn't help with de-confusing the host and path. I don't feel strongly about this.